<a href="https://colab.research.google.com/github/lordbigot/UTS_ML2019_ID13191655/blob/master/A3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

https://colab.research.google.com/drive/19GO4vG1Df6jkWbBOs9lrxm8SWhY6AJ56

https://github.com/lordbigot/UTS_ML2019_ID13191655/blob/master/A3.ipynb

# The Multi-armed Bandit Problem

## Introduction

Consider a machine with three arms. When the first arm is pulled, the machine returns \$5. When the second arm is pulled, the machine returns nothing. When the third arm is pulled, the machine has a 10% chance of returning \$1000 dollars. If the arms are pulled indefinitely, the third arm will produce the greatest reward. However, an observer can only learn the nature of this machine by pulling its arms. To get a complete picture, they are likely to start by pulling each arm once. In this case, there is a 90% chance that they will recieve no money from the third arm. When this occurs, they cannot distinguish between the second and third arms. From their perspective, it is possible neither arm will produce any reward.

This thought experiment brings me to a preliminary algorithm. First, each arm is pulled once. Then, the arms are sorted by their overall output. After this point, for as long as the machine can continue running, the arm with the greatest average output is pulled until the number of times it has been pulled exceeds n^2, where n is the number of times the arm with the lowest average output has been pulled. Each other arm is pulled, in series of highest to lowest average, until the number of times it has been pulled exceeds n. This algorithm will quickly scale, pulling the best-performing lever frequently, but continuing to pull the other levers as the risk persists that an otherwise poor-performing lever is the optimal solution.

## First steps

Consider a machine with 12 levers. You will only be able to pull a lever a number of times between 2 and 12. You are unaware of how many pulls you will recieve. Assume, each lever returns a constant reward, but the expected reward is unknown to you. In this case, it is apparent that attempting to pull as many unknown levers as possible will produce middling results, as will pulling one lever as many times as possible. An optimal solution may be to pull k levers, and then pull the lever that produced the highest reward. For all number of pulls n between 2 and 12, the ideal number for k sits between 2 and (n+1)/2 inclusive. If the distribution of lever rewards across the machine has a strong left skew, pulling the lever twice is ideal, as it is where the lowest values of lever are least commonly pulled. This is because there are the least chances of testing a low-valued lever, but if one is found, another lever can be used. Keep in mind that the algorithm has no way of evaluating if the reward is below normal until a second lever is pulled. If the distribution has a strong right skew, pulling the lever (n+1)/2 times is ideal, as it is where the highest values are most commonly pulled. This is because the best product from splitting a number into two numbers and then multiplying them together is a square. If n = 8, and k = 5, there are 5 chances of pulling the best lever, and if it is found, it will be pulled 3 more times, for a total of 4 times, and a product of 20 times across all possible cases. If n = 8, and k = 7, there are 7 chances of pulling the best lever. However, if it is found, it will be only be pulled one more time, for a total of 2 times. This produces the comparably poor product of 14 pulls across all possible cases. An algorithm to identify the proportions of possible pulls is included below.

In [0]:
def list_permutations(n, target_index, current_index):
    if current_index == 0:
        new_list = []
        for i in range(n):
            new_list.append([i])
        return new_list
    list_below = list_permutations(n, target_index, current_index - 1)
    new_list = []
    for row in list_below:
        new_elements = []
        for i in range(n):
            found_element = False
            for element in row:
                found_element = found_element or element == i
            if not found_element:
                new_elements.append([i])
        for element in new_elements:
            new_list.append(row + element)
    return new_list


def complete_list(n, data):
    missing_rows = n - len(data[0])
    for row in data:
        maximum_class = 0
        for element in row:
            maximum_class = max(maximum_class, element)
        for i in range(missing_rows):
            row.append(maximum_class)
    return data


def frequency_analysis(n, data):
    new_list = []
    for i in range(n):
        new_list.append(0)
    for row in data:
        for element in row:
            new_list[element] += 1
    divisor = len(data) * n
    for i in range(n):
        new_list[i] /= divisor
    return new_list


def frequency_comparison(n):
    print("n:", n)
    for i in range(n):
        print("k:", i+1)
        print(frequency_analysis(n, complete_list(n, list_permutations(n, i, i)))

Referring to our earlier problem, for all cases 3 <= n <= 12; k = 2 is a good solution. However, for the majority of values, k = 3 is likely to be a more rounded solution, less dependent on the assumption of a left skew. In this thought experiment, nothing is gained from pulling an apparently good lever before a new one is discovered. However, this will not always be the case. If we now assume that each lever has a variable result, then it may be useful to gather additional information on apparently good levers before moving onto other levers. In the short-term, this may increase overall output, and in the long-term, it will be insignificant, justified by the increased certainty of the effect of the lever. 

After examining this experiment, I am drawn to amend my earlier algorithm. This algorithm now has two processes. The overall process will pull two new levers, call the subprocess three times, pull a new lever, call the subprocess three times, pull another lever, and so on until it runs out new levers, at which point it continually calls the subprocess. The subprocess, when called, will sort the known levers from greatest average output to lowest average output. It will pull the first lever if the number of times it has been pulled is less than (x+1)^2, where x is the number of times the last lever has been pulled. Otherwise, it will pull the next following lever that has been pulled less than x+1 times. This algorithm will effectively produce the same results as the last algorithm if every lever can be pulled, but will produce above-average results faster than previously possible. This algorithm is encoded below.

In [0]:
def partition_levers(levers, output_sums, output_instances, low, high):
    i = (low - 1)
    pivot = output_sums[high] / output_instances[high]
    for j in range(low, high):
        if output_sums[j] / output_instances[j] >= pivot:
            i += 1
            levers[i], levers[j] = levers[j], levers[i]
            output_sums[i], output_sums[j] = output_sums[j], output_sums[i]
            output_instances[i], output_instances[j] = output_instances[j], output_instances[i]
    levers[i+1], levers[high] = levers[high], levers[i+1]
    output_sums[i+1], output_sums[high] = output_sums[high], output_sums[i+1]
    output_instances[i+1], output_instances[high] = output_instances[high], output_instances[i+1]
    return (i+1)


def quick_lever_sort(levers, output_sums, output_instances, low, high):
    if low < high:
        pi = partition_levers(levers, output_sums, output_instances, low, high)
        quick_lever_sort(levers, output_sums, output_instances, low, pi-1)
        quick_lever_sort(levers, output_sums, output_instances, pi+1, high)


def pull_known_lever(levers, output_sums, output_instances):
    quick_lever_sort(levers, output_sums, output_instances, 0, len(output_sums) - 1)
    x = output_instances[len(output_instances) - 1]
    if output_instances[0] < pow(x + 1, 2):
        output_sums[0] += pull_lever(levers[0])
        output_instances[0] += 1
        return
    current_lever = 1
    while current_lever < len(output_sums):
        if output_instances[current_lever] < x + 1:
            output_sums[current_lever] += pull_lever(levers[current_lever])
            output_instances[current_lever] += 1
            return
        current_lever += 1


def algorithm(levers):
    output_sums = []
    output_instances = []
    output_sums.append(pull_lever(levers[0]))
    output_instances.append(1)
    current_lever = 1
    while current_lever < len(levers):
        output_sums.append(pull_lever(levers[current_lever]))
        output_instances.append(1)
        for i in range(3):
            pull_known_lever(levers, output_sums, output_instances)
        current_lever += 1
    for i in range(100):
        pull_known_lever(levers, output_sums, output_instances)

## Algorithm Specialisation

This problem is well known in probability theory as the multi-armed bandit problem. However, most solutions assume four conditions that we are not given:

(1) The results from each lever are based on a continuous probability curve.

(2) The curve is without a significant skew.

(3) The number of arms is small.

(4) The amount of time it takes to test every arm is not worth considering, in comparison to the benefit of a preliminary examination of all variables.

These assumptions are useful, but not comprehensive. If I wish to address the problem I have been given, then solutions that make broad assumptions are undesirable. However, there is information that can be learned. A generic rule, Hoeffding's Inequality, can be used as a representation of all curves. It is a calculation for the maximum distance the true mean can vary from the real mean, within a certain confidence level, given the number of samples. By rearranging this equation, we can calculate a 95% confidence level around each lever. If the upper bound of this region is less than the lower bound of the best lever, than the current lever is no longer worth investigation. A solution to check the adjusted formula is included below:

In [0]:
def hoeffding_maximum_confident_displacement(number_of_samples, given_range):
    return 1.9206455826 * given_range / math.sqrt(number_of_samples)


def hoeffding_comparison(high_number_of_samples, high_mean, low_number_of_samples, low_mean, given_range):
    return low_mean + hoeffding_maximum_confident_displacement(low_number_of_samples, given_range) > high_mean - hoeffding_maximum_confident_displacement(high_number_of_samples, given_range)

## Standards and Finalisation

An important standard used in the evaluation of solutions to the multi-armed bandit problem is cumulative regret. Cumulative regret is the difference between the estimated mean of the best lever and the estimated mean of each other lever, multiplied by the times the other lever is used.

However, it produces an incomplete evaluation. Consider an algorithm that picked one lever and pulled it indefinitely. It would never gain enough information to regret a decision. However, it would be unlikely to select the optimal lever. If every lever gets pulled only once, the regret increases, and the chance of settling upon the optimal lever increases. The challenge is to minimise regret and maximise chance of finding the ideal lever.

As a seperate factor, a balance should be made between long term and short term cumulative regret. For good long term results, the best test that can be done is trying each lever in order, until the lever is confirmed to be no longer useful. For better short term results, testing of levers unlikely to return high values should be delayed, or stopped entirely.

Now that the components have been isolated, it may be worth re-examining some variables. My earlier rate of testing 1 new lever every 4 pulls was based on the assumption that the array was not skewed. However, given that we are testing gambling machines, the existence of a right skew is likely, as the existence of high jackpots is often used. As a result, I will change the rate to 1 in 3 pulls. In addition, this may decrease the long term cumulative regret, as good levers later in the set can be found before a significant investment is made in other, earlier levers.

Another factor worth reevaluating is the scaling proportions of the best and worst levers. The existing practice of the algorithm, using squares to ensure that pulling of lesser levers drops off, has drawbacks. To minimise long-term regret, the algorithm would function best by pulling each lever in uniform proportions. To minimise short-term regret, the algorithm would function best by pulling the favored lever repeatedly. To compromise, and perform well uniformly, it should make many pulls of the best performing lever, but also pull other, milder performing levers. In an attempt to be as fair as possible, I will build it so that it maintains a constant ratio of pulling lever n to pulling lever n+1, where the levers are ordered.

Factoring all of this into consideration, I have produced the following algorithm:

In [0]:
import math

def partition_levers(levers, output_sums, output_instances, low, high):
    i = (low - 1)
    pivot = output_sums[high] / output_instances[high]
    for j in range(low, high):
        if output_sums[j] / output_instances[j] >= pivot:
            i += 1
            levers[i], levers[j] = levers[j], levers[i]
            output_sums[i], output_sums[j] = output_sums[j], output_sums[i]
            output_instances[i], output_instances[j] = output_instances[j], output_instances[i]
    levers[i+1], levers[high] = levers[high], levers[i+1]
    output_sums[i+1], output_sums[high] = output_sums[high], output_sums[i+1]
    output_instances[i+1], output_instances[high] = output_instances[high], output_instances[i+1]
    return (i+1)


def quick_lever_sort(levers, output_sums, output_instances, low, high):
    if low < high:
        pi = partition_levers(levers, output_sums, output_instances, low, high)
        quick_lever_sort(levers, output_sums, output_instances, low, pi-1)
        quick_lever_sort(levers, output_sums, output_instances, pi+1, high)
        
        
def hoeffding_maximum_confident_displacement(number_of_samples, given_range):
    return 1.9206455826 * given_range / math.sqrt(number_of_samples)


def hoeffding_comparison(high_number_of_samples, high_mean, low_number_of_samples, low_mean, given_range):
    return low_mean + hoeffding_maximum_confident_displacement(low_number_of_samples, given_range) > high_mean - hoeffding_maximum_confident_displacement(high_number_of_samples, given_range)


def pull_known_lever(levers, output_sums, output_instances, given_range):
    quick_lever_sort(levers, output_sums, output_instances, 0, len(output_sums) - 1)
    known_levers = len(output_instances)
    minimum_lever_exponential_base = pow(output_instances[0], 1 / (2.0*(known_levers-1)))
    minimum_lever_exponential_base_index = 0
    for i in range(1, known_levers):
        if hoeffding_comparison(output_instances[0], output_sums[0] / output_instances[0], output_instances[i], output_sums[i] / output_instances[i], given_range):
            lever_exponential_base = pow(output_instances[i], 1 / (2.0*(known_levers-1)-i))
            if lever_exponential_base < minimum_lever_exponential_base:
                minimum_lever_exponential_base = lever_exponential_base
                minimum_lever_exponential_base_index = i
    result = pull_lever(levers[minimum_lever_exponential_base_index])
    output_sums[minimum_lever_exponential_base_index] += pull_lever(levers[minimum_lever_exponential_base_index])
    output_instances[minimum_lever_exponential_base_index] += 1
    return result


def algorithm(levers, pulls):
    output_sums = []
    output_instances = []
    minimum_value = pull_lever(levers[0])
    maximum_value = minimum_value
    intermediate_range = 0.001
    output_sums.append(minimum_value)
    output_instances.append(1)
    count = 1
    current_lever = 1
    while current_lever < len(levers):
        recent_result = pull_lever(levers[current_lever])
        count += 1
        if recent_result < minimum_value:
            minimum_value = recent_result
            intermediate_range = maximum_value - minimum_value
        elif recent_result > maximum_value:
            maximum_value = recent_result
            intermediate_range = maximum_value - minimum_value
        output_sums.append(recent_result)
        output_instances.append(1)
        for i in range(2):
            recent_result = pull_known_lever(levers, output_sums, output_instances, intermediate_range)
            count += 1
            if recent_result < minimum_value:
                minimum_value = recent_result
                intermediate_range = maximum_value - minimum_value
            elif recent_result > maximum_value:
                maximum_value = recent_result
                intermediate_range = maximum_value - minimum_value
        current_lever += 1
    for i in range(count, pulls):
        recent_result = pull_known_lever(levers, output_sums, output_instances, intermediate_range)
        if recent_result < minimum_value:
            minimum_value = recent_result
            intermediate_range = maximum_value - minimum_value
        elif recent_result > maximum_value:
            maximum_value = recent_result
            intermediate_range = maximum_value - minimum_value
    result_sum = 0
    for current_sum in output_sums:
        result_sum += current_sum
    return result_sum

## Testing

For the purposes of testing, a comprehensive lever capable of replicating a variety of test conditions is desirable.

The below types of levers are:

0: The lever has a given chance of returning 1, otherwise returns 0. This is a special case of the following lever

1: The lever has a given chance of returning a given value, a given chance of returning another value, and so on

2: The lever returns values along to a gaussian distribution.

In [0]:
import random

def pull_lever(lever):
  if lever[0] == 0:
    if random.random() < lever[1]:
      return 1
    else:
      return 0
  elif lever[0] == 1:
    index = 1
    key = random.random()
    while index + 1 < len(lever):
      if key < lever[index]:
        return lever[index + 1]
      index += 2
    return lever[index]
  elif lever[0] == 2:
    return random.gauss(lever[1], lever[2])
  else:
    return 0

As an initial test, I will try using a series of 10 levers, each with a different chance of returning a consistent reward, from 0% to 90%.

100 pulls is a good middle-of-the-range number, where information is gathered, but not

Given 100 pulls, a solution with perfect information would on average produce 90 points. A solution with no information would on average produce 45 points.

In [88]:
test_length = 100
number_of_pulls = 10000

# in order
point_sum = 0
for i in range(test_length):
    levers = [[0,0.0],[0,0.1],[0,0.2],[0,0.3],[0,0.4],[0,0.5],[0,0.6],[0,0.7],[0,0.8],[0,0.9]]
    point_sum += algorithm(levers,number_of_pulls)
print(point_sum/test_length)

# in reverse order
point_sum = 0
for i in range(test_length):
    levers = [[0,0.9],[0,0.8],[0,0.7],[0,0.6],[0,0.5],[0,0.4],[0,0.3],[0,0.2],[0,0.1],[0,0.0]]
    point_sum += algorithm(levers,number_of_pulls)
print(point_sum/test_length)

# in unambiguous order
point_sum = 0
for i in range(test_length):
    levers = [[0,0.9],[0,0.0],[0,0.1],[0,0.2],[0,0.3],[0,0.4],[0,0.5],[0,0.6],[0,0.7],[0,0.8]]
    point_sum += algorithm(levers,number_of_pulls)
print(point_sum/test_length)

# in random orders
point_sum = 0
for i in range(test_length):
    random.shuffle(levers)
    point_sum += algorithm(levers,number_of_pulls)
print(point_sum/test_length)

8705.77
8695.66
8704.76
8702.08


I was able to accumulate the following averages, given 100 lever pulls:

in order: 64.3968

in reverse order: 64.2870

in unambiguous order: 64.3507

in random order: 64.3114

This demonstrates that the grouping of levers is mostly insignificant in the overall outcome, which is definitively influenced by the information gathered.

In addition, I have data on the cumulative regret of other algorithms performing the same test.

If this algorithm gained perfect information, its cumulative regret at 2500 time steps would be equal to 2250 minus the average point total. This is 1992.94, leaving the regret at 247.83, worse than UCB1 (Weng 2018). The regret at 10000 time steps is 302.34, better than UCB1.

As a second test, I will use a massive jackpot, occurring at only 20% frequency, and compare it to a constant reward function.

In each case:

a solution with perfect information should have an average return of 20000

a solution favoring the wrong choice should have an average return of 10000

a solution with no information (ignoring decoys), should have an average return of 15000

a solution with no information (including decoys), should have an average return of 3000

In [73]:
test_trials = 10000
number_of_pulls = 100

# jackpot is better than constant
levers = [[1,1.0,100],[1,0.2,1000,0]]
point_sum = 0
for i in range(test_trials):
    random.shuffle(levers)
    point_sum += algorithm(levers,number_of_pulls)
print(point_sum/test_trials)

# jackpot is better than constant, plus decoys
levers = [[1,1.0,100],[1,0.2,1000,0],[0,0.0],[0,0.0],[0,0.0],[0,0.0],[0,0.0],[0,0.0],[0,0.0],[0,0.0]]
point_sum = 0
for i in range(test_trials):
    random.shuffle(levers)
    point_sum += algorithm(levers,number_of_pulls)
print(point_sum/test_trials)

# constant is better than jackpot
levers = [[1,1.0,200],[1,0.2,500,0]]
point_sum = 0
for i in range(test_trials):
    random.shuffle(levers)
    point_sum += algorithm(levers,number_of_pulls)
print(point_sum/test_trials)

# constant is better than jackpot, plus decoys
levers = [[1,1.0,200],[1,0.2,500,0],[0,0.0],[0,0.0],[0,0.0],[0,0.0],[0,0.0],[0,0.0],[0,0.0],[0,0.0]]
point_sum = 0
for i in range(test_trials):
    random.shuffle(levers)
    point_sum += algorithm(levers,number_of_pulls)
print(point_sum/test_trials)

16387.6
4196.9779
17415.44
4615.6946


I found the following averages under 100 pulls:

Jackpot is better than constant: 16315.06

Jackpot is better than constant (plus decoys): 4192.5923

Constant is better than the jackpot: 17415.44

Constant is better than jackpot (plus decoys): 4615.6946

However, under the conditions, 100 pulls may not have given the algorithm enough information to demonstrate the capabilities of the function, so I also tested the same scenario on 1000 pulls, and divided the results by 10 (for the purposes of direct comparison)

Jackpot is better than constant: 18269.384

Jackpot is better than constant (plus decoys): 7154.03009

Constant is better than jackpot: 18529.368

Constant is better than jackpot (plus decoys): 9674.27925

Having proven the efficacy of this algorithm with discrete values, I would like to confirm that it is simultaneously suitable for continuous functions. I will test a series of continuous and discrete levers.

The best case scenario for this algorithm is 10000.
If the algorithm does not work, the result will be closer to 6767.

In [84]:
test_trials = 1000
number_of_pulls = 100

# continuous is better than discrete
levers = [[2,100,10],[2,70,100],[2,70,1],[1,0.5,150,0],[1,0.1,100,0.5,110,0.9,90,0],[0,1.0]]
point_sum = 0
for i in range(test_trials):
    random.shuffle(levers)
    point_sum += algorithm(levers,number_of_pulls)
print(point_sum/test_trials)

# discrete is better than continuous
levers = [[2,75,10],[2,70,100],[2,70,1],[1,0.5,200,0],[1,0.1,100,0.5,110,0.9,90,0],[0,1.0]]
point_sum = 0
for i in range(test_trials):
    random.shuffle(levers)
    point_sum += algorithm(levers,number_of_pulls)
print(point_sum/test_trials)

7816.88411779411
7750.310547243608


The results in 100 pulls are approximately:

continuous normal curve is best: 7816.88

discrete dichotomy is best: 7750.310547243608

## Conclusion

In summary, The Multi-armed bandit has a lot of factors, and a lot of possible solutions.

Some solutions are specialised for particular distributions, and the majority are optimised for maximising output at high values. The standard approach to the exploration/exploitation trade-off favors exploration at low values, and expoitation at high values.

I believe the algorithm I have produced is genuinely good for very small numbers of lever pulls, despite the compromise in early information gain. However, if I was to look for a true algorithm for every circumstance, I would likely pick a form of Thompson Sampling (Chapelle et al. 2011). Amongst other options, it makes a comparatively small number of assumptions.



## Bibliography

Weng, L. 2018, 'The Multi-Armed Bandit Problem and Its Solutions', *Lil'Log*, weblog, github, Indian Ocean, 23 January, viewed 8 October 2019, <https://lilianweng.github.io/lil-log/2018/01/23/the-multi-armed-bandit-problem-and-its-solutions.html>.

Chapelle, O., & Lihong L. 2011, 'An empirical evaluation of thompson sampling.', *Advances in neural information processing systems*

Martín, M., Jiménez-Martín A., & Mateos A. 2018, 'Possibilistic reward methods for the multi-armed bandit problem'
*Neurocomputing*, vol. 310, pp. 201-212.